Src/Support/Google.Apis.Core/Http/HttpClientFromMessageHandlerFactory.cs

```diff
@@ -0,0 +1,198 @@
+        /// message handlers that will be used when creating the <see cref="ConfigurableHttpClient"/>.
+        /// </summary>
+        /// <remarks>
+        /// The <see cref="HttpMessageHandler"/> obtained from the factory won't be disposed
+        /// when the created <see cref="ConfigurableHttpClient"/> is. This allows calling code
+        /// to control the handlers' lifetime and so they can possibly be reused.
+        /// This may be a requirement for using System.Net.Http.IHttpMessageHandler. See
+        /// https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests
+        /// for information on why to use System.Net.Http.IHttpMessageHandler.
+        /// </remarks>
+        public HttpClientFromMessageHandlerFactory(Func<HttpMessageHandlerOptions, ConfiguredHttpMessageHandler> httpMessageHandlerFactory) =>
+            _httpMessageHandlerFactory = httpMessageHandlerFactory.ThrowIfNull(nameof(httpMessageHandlerFactory));
+
+        /// <inheritdoc/>
+        public ConfigurableHttpClient CreateHttpClient(CreateHttpClientArgs args)
+        {
+            var handler = CreateHandler(args);
+            var configurableHandler = new ConfigurableMessageHandler(handler)
+            {
+                ApplicationName = args.ApplicationName,
+                GoogleApiClientHeader = args.GoogleApiClientHeader
+            };
+
+            // We always set not to dispose the inner handler, it's not created
+            // by us so we don't control it's lifetime.
+            var client = new ConfigurableHttpClient(configurableHandler, false);
+            foreach (var initializer in args.Initializers)
+            {
+                initializer.Initialize(client);
+            }
+
+            return client;
+        }
+
+        private HttpMessageHandler CreateHandler(CreateHttpClientArgs args)
+        {
+            // We need to handle three situations in order to intercept uncompressed data where necessary
+            // while using the built-in decompression where possible.
+            // - No compression requested
+            // - Compression requested but not supported by the injected message handler
+            //   (easy; just GzipDeflateHandler on top of an interceptor on top of HttpMessagetHandler)
+            // - Compression requested and supported by the injected message handler
+            //   (complex: create two different handlers and decide which to use on a per-request basis)
+
+            // First request a message handler that may perform decompression or not depending on whether
+            // GZip is enabled.
+            var effectiveOptions = args.GZipEnabled ? AllowDecompressionOptions : DisallowDecompressionOptions;
+            var configuredMessageHandler = effectiveOptions.CheckItMatches(_httpMessageHandlerFactory(effectiveOptions));
+
+            if (!args.GZipEnabled)
```

Comment:
It might be clearer to use effectiveOptions.MayPerformDecompression here, as it's closer to describing the handler's behavior.